### PR TITLE
consensus/cw_calculator: add as approvers txs which has approvees pre…

### PR DESCRIPTION
**Pull request by tsvisabo**

### CODE REVIEW

File: ciri/consensus/tip_selection/cw_rating_calculator/cw_rating_dfs_impl.c [hunk 1]
```diff
@@ -27,53 +27,68 @@ static retcode_t cw_rating_dfs_do_dfs_from_db(tangle_t *const tangle, flex_trit_
                                               int64_t subtangle_before_timestamp) {
   hash_to_indexed_hash_set_entry_t *curr_tx = NULL;
   retcode_t ret = RC_OK;
-  iota_stor_pack_t pack;
+  iota_stor_pack_t apprvoers_pack;
```
Comment (thibault-martinez) [inline ciri/consensus/tip_selection/cw_rating_calculator/cw_rating_dfs_impl.c]: Typo

File: ciri/consensus/tip_selection/cw_rating_calculator/cw_rating_dfs_impl.c [hunk 2]
```diff
@@ -27,53 +27,68 @@ static retcode_t cw_rating_dfs_do_dfs_from_db(tangle_t *const tangle, flex_trit_
...
-          goto done;
+      while (approvers_pack.num_loaded > 0) {
+        curr_tx_hash = ((flex_trit_t *)approvers_pack.models[--approvers_pack.num_loaded]);
+
+        ERR_BIND_GOTO(iota_tangle_transaction_load_partial(tangle, curr_tx_hash, &transaction_pack,
+                                                           PARTIAL_TX_MODEL_ESSENCE_ATTACHMENT_METADATA),
+                      ret, done)
+        hash_pack_reset(&approvees_pack);
+
+        if (transaction_trunk(&tx) == transaction_branch(&tx)) {
```
Comment (thibault-martinez) [inline ciri/consensus/tip_selection/cw_rating_calculator/cw_rating_dfs_impl.c]: Use `memcmp`

File: ciri/consensus/tip_selection/cw_rating_calculator/cw_rating_dfs_impl.c [hunk 3]
```diff
@@ -27,53 +27,68 @@ static retcode_t cw_rating_dfs_do_dfs_from_db(tangle_t *const tangle, flex_trit_
...
+          hash243_queue_push(&approvees, transaction_branch(&tx));
+        } else {
+          hash243_queue_push(&approvees, transaction_branch(&tx));
+          hash243_queue_push(&approvees, transaction_trunk(&tx));
         }
-        if ((ret = hash243_set_add(&curr_tx->approvers, curr_tx_hash)) != RC_OK) {
-          goto done;
+
+        ERR_BIND_GOTO(iota_tangle_transaction_find(tangle, NULL, NULL, NULL, approvees, &approvees_pack), ret, done);
```
Comment (thibault-martinez) [inline ciri/consensus/tip_selection/cw_rating_calculator/cw_rating_dfs_impl.c]: This one was created for the API call `findTransactions` and is *really* slow, I'm not sure if we should use it in CW.

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

File: ciri/consensus/tip_selection/cw_rating_calculator/cw_rating_dfs_impl.c (hunk 2 above)
Comment (tsvisabo) [reply ciri/consensus/tip_selection/cw_rating_calculator/cw_rating_dfs_impl.c]: whoops

Comment (thibault-martinez) [reply ciri/consensus/tip_selection/cw_rating_calculator/cw_rating_dfs_impl.c]: C++ habits :trollface:

File: ciri/consensus/tip_selection/cw_rating_calculator/cw_rating_dfs_impl.c [hunk 4]
```diff
@@ -28,55 +28,79 @@ static retcode_t cw_rating_dfs_do_dfs_from_db(tangle_t *const tangle, flex_trit_
...
   uint64_t start_timestamp, end_timestamp;
 
   *subtangle_size = 0;
 
   start_timestamp = current_timestamp_ms();
 
-  if ((ret = hash_pack_init(&pack, 10)) != RC_OK) {
-    goto done;
-  }
+  DECLARE_PACK_SINGLE_TX(tx, tx_ptr, transaction_pack);
```
Comment (thibault-martinez) [inline ciri/consensus/tip_selection/cw_rating_calculator/cw_rating_dfs_impl.c]: Declare at the top

File: ciri/consensus/tip_selection/cw_rating_calculator/cw_rating_dfs_impl.c [hunk 5]
```diff
@@ -28,55 +28,79 @@ static retcode_t cw_rating_dfs_do_dfs_from_db(tangle_t *const tangle, flex_trit_
...
 
-  if ((ret = hash243_stack_push(&stack, entry_point)) != RC_OK) {
-    goto done;
-  }
+  *subtangle_size = 0;
```
Comment (thibault-martinez) [inline ciri/consensus/tip_selection/cw_rating_calculator/cw_rating_dfs_impl.c]: Already done line 41

File: ciri/consensus/tip_selection/cw_rating_calculator/cw_rating_dfs_impl.c [hunk 6]
```diff
@@ -28,55 +28,79 @@ static retcode_t cw_rating_dfs_do_dfs_from_db(tangle_t *const tangle, flex_trit_
...
+                                                           PARTIAL_TX_MODEL_ESSENCE_ATTACHMENT_METADATA),
+                      ret, done)
+
+        ERR_BIND_GOTO(hash243_stack_push(&stack, curr_tx_hash), ret, done);
+        both_approvees_exist = true;
+        if (!transaction_solid(&tx)) {
+          hash243_set_add(&approvees, transaction_branch(&tx));
+          hash243_set_add(&approvees, transaction_trunk(&tx));
+
+          HASH_SET_ITER(approvees, iter, tmp) {
```
Comment (thibault-martinez) [inline ciri/consensus/tip_selection/cw_rating_calculator/cw_rating_dfs_impl.c]: Do we really need to have the overhead of a set for only two approvees ? Can't we just do two `exist` ?

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

**Review by thibault-martinez — DISMISSED** [review]

File: ciri/consensus/tip_selection/cw_rating_calculator/cw_rating_dfs_impl.c [hunk 7]
```diff
@@ -28,46 +28,59 @@ static retcode_t cw_rating_dfs_do_dfs_from_db(tangle_t *const tangle, flex_trit_
                                               int64_t subtangle_before_timestamp) {
   hash_to_indexed_hash_set_entry_t *curr_tx = NULL;
   retcode_t ret = RC_OK;
-  iota_stor_pack_t pack;
+  iota_stor_pack_t approvers_pack;
   hash243_stack_t stack = NULL;
   flex_trit_t *curr_tx_hash = NULL;
+  bool trunk_exist;
```
Comment (thibault-martinez) [inline ciri/consensus/tip_selection/cw_rating_calculator/cw_rating_dfs_impl.c]:
```suggestion
  bool trunk_exist = false;
```

File: ciri/consensus/tip_selection/cw_rating_calculator/cw_rating_dfs_impl.c [hunk 8]
```diff
@@ -28,46 +28,59 @@ static retcode_t cw_rating_dfs_do_dfs_from_db(tangle_t *const tangle, flex_trit_
...
+  bool branch_exist;
```
Comment (thibault-martinez) [inline ciri/consensus/tip_selection/cw_rating_calculator/cw_rating_dfs_impl.c]:
```suggestion
  bool branch_exist = false;
```

**Review by thibault-martinez — APPROVED** [review]

**Pull request merged**